File: tictactoe.py
class TicTacToe():
    def __init__(self, initial='X'):
        self.start = initial
        self.board = [' ' for _ in range(9)]
        self.current_winner = None
    
    # print a big board with numbers on the right for reference
    def print_board(self):
        for i in range(3):
            print(' | '.join(self.board[i*3:i*3+3]) + '   ' + ' | '.join(str(x) for x in range(i*3+1, i*3+4)))
            if i < 2:
                print('-'*10)
    
    def board_full(self):
        return ' ' not in self.board
    
    def check_winner(self):
        b = self.board
        # horizontal
        for i in range(3):
            if b[i*3] == b[i*3+1] == b[i*3+2] and b[i*3] != ' ':
                return b[i*3]
        # vertical
        for i in range(3):
            if b[i] == b[i+3] == b[i+6] and b[i] != ' ':
                return b[i]
        # diagonal
        if ( b[0] == b[4] == b[8] or b[6] == b[4] == b[2] ) and b[4] != ' ':
            return b[4]
    
    def get_move(self, player): # get move from player

        if player == 'X': # human
            while True: # loop until valid input
                try:
                    play_num = int(input("Enter your move (1-9): "))
                    if play_num in range(1, 10) and self.board[play_num-1] == ' ':
                        return play_num
                    else:
                        raise ValueError
                except ValueError:
                    print("Invalid move.")

        else: # bot
            return self.bot_minimax()
    
    def bot_random(self): # random bot
        import random
        while True:
            play_num = random.randint(1, 9)
            if self.board[play_num-1] == ' ':
                return play_num
    

    def minimax(self, board, depth, is_maximizing): # minimax algorithm
        scores = {'X': -1, 'O': 1, 'tie': 0} # scores for each player
        winner = self.check_winner() # check if game is over
        if winner:
            return scores[winner]
        if self.board_full():
            return scores['tie']

        if is_maximizing: # bot's turn
            best_score = float('-inf')
            for i in range(9):
                if board[i] == ' ':
                    board[i] = 'O'
                    if self.check_winner() == 'O':
                        board[i] = ' '
                        return scores['O']
                    score = self.minimax(board, depth + 1, False)
                    board[i] = ' '
                    best_score = max(score, best_score)
            return best_score
        
        else: # player's turn
            best_score = float('inf')
            for i in range(9):
                if board[i] == ' ':
                    board[i] = 'X'
                    if self.check_winner() == 'X':
                        board[i] = ' '
                        return scores['X']
                    score = self.minimax(board, depth + 1, True)
                    board[i] = ' '
                    best_score = min(score, best_score)
            return best_score
    
    def bot_minimax(self): # minimax bot
        best_score = float('-inf')
        best_move = None
        for i in range(9): # loop through all possible moves
            if self.board[i] == ' ': # check if move is valid
                self.board[i] = 'O' # make move
                if self.check_winner() == 'O':  # Check for immediate win
                    self.board[i] = ' '
                    return i + 1
                score = self.minimax(self.board, 0, False) # get score for move
                self.board[i] = ' ' # undo move
                if score > best_score: # update best move
                    best_score = score
                    best_move = i + 1
        return best_move

    def play_game(self): # main loop
        player = 'X' if self.start == 'X' else 'O'
        while not self.board_full():
            self.print_board()
            print("player",player)
            play_num = self.get_move(player)
            self.board[play_num-1] = player
            player = 'O' if player == 'X' else 'X'
            winner = self.check_winner()
            if winner:
                break
        self.print_board()
        if winner:
            print(winner, "wins!")
        else:
            print("It's a draw!")

    
def main():
    game = TicTacToe(initial='O') # initial player
    game.play_game()

if __name__ == '__main__':
    main()

File: tictactoe.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// create and return an initialized board
char* create_board() {
    char* board = (char*)malloc(9 * sizeof(char));
    for (int i = 0; i < 9; i++) {
        board[i] = ' ';
    }
    return board;
}

// print the board, with a reference board on the right (1-9)
void print_board(char* board) {
    for (int i = 0; i < 9; i += 3) {
        printf(" %c | %c | %c    %d | %d | %d\n", board[i], board[i+1], board[i+2], i+1, i+2, i+3);
        if (i < 6) {
            printf("---|---|---  ---|---|---\n");
        }
    }
}

// check if the board is full
int is_full(char* board) {
    for (int i = 0; i < 9; i++) {
        if (board[i] == ' ') {
            return 0;
        }
    }
    return 1;
}

// check if the player has won
int has_won(char* board, char player) {
    for (int i = 0; i < 3; i++) {
        if (board[i] == player && board[i+3] == player && board[i+6] == player) {
            return 1;
        }
        if (board[3*i] == player && board[3*i+1] == player && board[3*i+2] == player) {
            return 1;
        }
    }
    if (board[0] == player && board[4] == player && board[8] == player) {
        return 1;
    }
    if (board[2] == player && board[4] == player && board[6] == player) {
        return 1;
    }
    return 0;
}

// get move input from player
int get_move(char* board) {
    int move;
    while (1) {
        printf("Enter your move (1-9): ");
        scanf("%d", &move);
        if (move >= 1 && move <= 9 && board[move-1] == ' ') {
            return move;
        }
        printf("Invalid move.\n");
    }
}

// bot random move
int bot_move(char* board) {
    int move;
    while (1) {
        move = rand() % 9;
        if (board[move-1] == ' ') {
            return move;
        }
    }
}

// minimax algorithm
int minimax(char* board, int depth, int is_maximizing) {
    if (has_won(board, 'X')) {
        return -10;
    }
    if (has_won(board, 'O')) {
        return 10;
    }
    if (is_full(board)) {
        return 0;
    }
    if (is_maximizing) { // bot's turn
        int best_score = -1000;
        for (int i = 0; i < 9; i++) {
            if (board[i] == ' ') {
                board[i] = 'O';
                int score = minimax(board, depth+1, 0);
                board[i] = ' ';
                if (score > best_score) {
                    best_score = score;
                }
            }
        }
        return best_score;
    } else { // player's turn
        int best_score = 1000;
        for (int i = 0; i < 9; i++) {
            if (board[i] == ' ') {
                board[i] = 'X';
                int score = minimax(board, depth+1, 1);
                board[i] = ' ';
                if (score < best_score) {
                    best_score = score;
                }
            }
        }
        return best_score;
    }
}

// bot minimax (don't forget to check the current move)
int bot_minimax(char* board) {
    int best_score = -1000;
    int best_move;
    for (int i = 0; i < 9; i++) {
        if (board[i] == ' ') {
            board[i] = 'O';
            if (has_won(board, 'O')) { // check winning move
                board[i] = ' ';
                return i+1;
            }
            int score = minimax(board, 0, 0);
            printf("Move %d: %d\n", i+1, score); // debug
            board[i] = ' ';
            if (score > best_score) {
                best_score = score;
                best_move = i+1;
            }
        }
    }
    return best_move;
}


// play the game
void play_game(char initial) {
    char* board = create_board();
    char player = initial;
    int move;
    while (1) {
        print_board(board);
        if (player == 'X') {
            move = get_move(board);
        } else {
            move = bot_minimax(board);
        }
        board[move-1] = player;
        if (has_won(board, player)) {
            print_board(board);
            printf("%c wins!\n", player);
            break;
        }
        if (is_full(board)) {
            print_board(board);
            printf("It's a draw!\n");
            break;
        }
        player = (player == 'X') ? 'O' : 'X';
    }
    free(board);
}


int main() {
    srand(time(NULL));
    play_game('O');
    return 0;
}